File: python/46.permutations.py
class Solution(object):
    permutations = []
    taken = {}

    def back_track(self, idx, nums, path):
        if len(path) == len(nums):
            self.permutations.append(path[:])

        for i in range(len(nums)):
            if self.taken[nums[i]]:
                continue
            self.taken[nums[i]] = True
            path.append(nums[i])
            self.back_track(i, nums, path)
            path.pop()
            self.taken[nums[i]] = False


    def solve1(self, nums):
        self.permutations = []
        self.taken = {i:False for i in nums}
        self.back_track(0, nums, [])
        return self.permutations

    def back_track2(self, idx, nums):
        if idx == len(nums):
           self.permutations.append(nums[:])
           return
        for i in range(idx, len(nums)):
           nums[idx], nums[i] = nums[i], nums[idx]
           self.back_track2(idx+1, nums)
           nums[idx], nums[i] = nums[i], nums[idx]

    def solve2(self, nums):
        self.permutations = []
        self.taken = {i:False for i in nums}
        self.back_track2(0, nums)
        return self.permutations

    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        return self.solve2(nums)

File: cpp/46.permutations.cpp
class Solution {
public:
    //marked 1 1 0 1 
    //       1 2 3 4
    //             l
    //path:  1 2 4
    //results=
    //         1 2 3 4
    void bk(int level, vector<int> &nums, vector<int> &path, vector<vector<int> > &result, vector<bool> &marked){
	 if(level == 0){
	     result.push_back(path);
	     return;
	 }
	 for(int i = 0 ;i < nums.size(); i++){
            if(marked[i])continue;
	    marked[i] = true;
	    path.push_back(nums[i]);
	    bk(level-1, nums, path, result, marked);
	    marked[i] = false;
	    path.pop_back();
	 }
    }    
    vector<vector<int> > solve(vector<int> &nums){
	  vector<vector<int> > res;
	  vector<int> path;
	  vector<bool> marked((int) nums.size());
	  bk(nums.size(), nums, path, res, marked);
	  return res;
    }
    vector<vector<int>> permute(vector<int>& nums) {
	  return solve(nums);
    }
};
